pkg/controller/common/annotation/controller_version.go

```diff
@@ -59,15 +43,28 @@ func UpdateControllerVersion(ctx context.Context, client k8s.Client, obj ctrlcli
 	}
 
 	annotations[ControllerVersionAnnotation] = version
-	err = accessor.SetAnnotations(obj, annotations)
-	if err != nil {
-		log.Error(err, "error updating controller version annotation", "namespace", namespace, "name", name, "kind", obj.GetObjectKind())
+	accessor := meta.NewAccessor()
+	if err := accessor.SetAnnotations(obj, annotations); err != nil {
+		log.Error(err, "error updating controller version annotation", "namespace", obj.GetNamespace(), "name", obj.GetName(), "kind", obj.GetObjectKind())
 		return err
 	}
-	log.V(1).Info("updating controller version annotation", "namespace", namespace, "name", name, "kind", obj.GetObjectKind())
+	log.V(1).Info("updating controller version annotation", "namespace", obj.GetNamespace(), "name", obj.GetName(), "kind", obj.GetObjectKind())
 	return client.Update(context.Background(), obj)
 }
 
+// CheckCompatibility determines if this controller is compatible with a given resource by examining the controller version annotation. It has no side effect and
+// can be used by auxiliary controllers to check if they can process a resource.
+// First boolean is true if the compatibility cannot be determined at the moment, mostly because the main controller did not reconcile the resource yet. In that case the
+// auxiliary controller must try again later.
+func CheckCompatibility(obj ctrlclient.Object, controllerVersion string) (requeue bool, supported bool, err error) {
```

Comment:
How about returning a specific error (e.g. `ErrNotReconciled`) if the resource is not reconciled yet?

Returning two booleans from this method feels weird to me.  Since the function is called `CheckCompatibility`, I would expect it to return a true/false result about the compatibility. Whether to requeue or not is a decision that the caller can make for themselves by inspecting the error.

Reply:
I think it does not feel like an error to me, also returning such a specific error somewhat reminds me the "coding by exception" anti-pattern.
I would also expect this function to be called by controllers, they would all have to check for this error. Given the purpose of the function I'm not sure there is a situation where we would just ignore it and return it to the caller.

I agree that retuning multiple values is not common, I usually avoid that. But in that case I would say it seems "acceptable" ?

Reply:
I would say that in Go checking for specific kinds of errors is a common pattern (they are not "exceptions" that cause an expensive stack unwinding like in Java). To me this function seems to have multiple responsibilities. The controllers will still need to check whether `requeue == true` which is no more different than `errors.Is(err, ErrNotReconciled)`. The latter feels more idiomatic.

Reply:
I removed the first boolean as it is the responsibility for the caller to watch for changes on the resource anyway.
I still feel like using an error while there is none (it is an expected situation) seems wrong, this what I'm referring to with the "coding by exception" anti-pattern.
It meets the current need and can be revisited later if needed.